crates/algokit_utils/src/clients/testnet_dispenser_client.rs

```diff
@@ -0,0 +1,259 @@
+
+/// The TestNet Dispenser API response when getting the current limit.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct DispenserLimitResponse {
+    /// The limit, in µAlgo, that you can currently fund.
+    pub amount: u64,
+}
+
+/// Errors that can occur when using the TestNet Dispenser API client.
+#[derive(Debug, Snafu)]
+pub enum DispenserError {
+    #[snafu(display(
+        "Can't init AlgoKit TestNet Dispenser API client because neither environment variable {DISPENSER_ACCESS_TOKEN_KEY} nor the authToken were provided"
+    ))]
+    MissingAuthToken,
+
+    #[snafu(display("Error processing dispenser API request: {message}"))]
+    ApiError { message: String },
+
+    #[snafu(display("HTTP request failed: {source}"))]
+    RequestError { source: reqwest::Error },
+
+    #[snafu(display("Failed to parse response: {source}"))]
+    ParseError { source: reqwest::Error },
+}
+
+#[derive(Deserialize)]
+struct ErrorResponse {
+    #[serde(rename = "code")]
+    code: Option<String>,
+    #[serde(rename = "message")]
+    message: Option<String>,
+}
+
+#[derive(Serialize)]
+struct FundRequest {
+    receiver: String,
+    amount: u64,
+    #[serde(rename = "assetID")]
+    asset_id: u64,
+}
+
+#[derive(Deserialize)]
+struct FundResponseDto {
+    #[serde(rename = "txID")]
+    tx_id: String,
+    amount: u64,
+}
+
+#[derive(Serialize)]
+struct RefundRequest {
+    #[serde(rename = "refundTransactionID")]
+    refund_transaction_id: String,
+}
+
+#[derive(Deserialize)]
+struct LimitResponseDto {
+    amount: u64,
+}
+
+static ALGO_ASSET_ID: u64 = 0;
+
+/// The parameters to construct a TestNet Dispenser API client.
+#[derive(Debug, Clone)]
+pub struct TestNetDispenserApiClientParams {
+    /// The authentication token, otherwise an `ALGOKIT_DISPENSER_ACCESS_TOKEN` environment variable
+    /// will be used. If neither is provided, an error is returned.
+    pub auth_token: Option<String>,
+    /// The request timeout in seconds. If not provided, a default value of 15 seconds is used.
+    pub request_timeout: Option<u64>,
+}
+
+/// `TestNetDispenserApiClient` is a client that provides methods to interact with the
+/// [Algorand TestNet Dispenser API](https://github.com/algorandfoundation/algokit/blob/main/docs/testnet_api.md).
+/// It allows you to fund an address with Algo, refund a transaction, and get the funding limit for the Algo asset.
+pub struct TestNetDispenserApiClient {
+    auth_token: String,
+    http_client: Client,
+}
+
+impl TestNetDispenserApiClient {
+    /// Create a new TestNet Dispenser API client.
+    ///
+    /// # Parameters
+    /// * `params` - Optional TestNetDispenserApiClientParams parameters.
+    ///
+    /// # Returns
+    /// A new `TestNetDispenserApiClient` instance or an error if the auth token cannot be determined.
+    pub fn new(params: Option<TestNetDispenserApiClientParams>) -> Result<Self, DispenserError> {
+        use std::env;
+
+        let auth_token = if let Some(ref p) = params {
```

Comment:
If a user passes Some(TestNetDispenserApiClientParams { auth_token: None, ... }), the environment variable will be ignored